subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ForkingGradleExecuter.java

```diff
@@ -179,20 +179,22 @@ public File getWorkingDir() {
     }
 
     @Override
-    public GradleHandle doStart() {
-        return createGradleHandle(getResultAssertion(), getDefaultCharacterEncoding(), new Factory<DefaultExecHandleBuilder>() {
+    public GradleHandle createGradleHandle() {
+        return createForkingGradleHandle(getResultAssertion(), getDefaultCharacterEncoding(), new Factory<DefaultExecHandleBuilder>() {
             public DefaultExecHandleBuilder create() {
                 return createExecHandleBuilder();
             }
         }).start();
     }
 
-    protected ForkingGradleHandle createGradleHandle(Action<ExecutionResult> resultAssertion, String encoding, Factory<? extends AbstractExecHandleBuilder> execHandleFactory) {
+    protected ForkingGradleHandle createForkingGradleHandle(Action<ExecutionResult> resultAssertion, String encoding, Factory<? extends AbstractExecHandleBuilder> execHandleFactory) {
         return new ForkingGradleHandle(getStdinPipe(), isUseDaemon(), resultAssertion, encoding, execHandleFactory, getDurationMeasurement());
     }
 
     protected ExecutionResult doRun() {
-        return start().waitForFinish();
```

Comment:
This was the fundamental problem.

start() asserted that there were no afterExecute actions, because they aren't supported for async execution. This meant that `gradleExecuter.afterExecute()` bombed with `ForkingGradleExecuter`.